File: app/src/main/java/com/example/a4/Adaptor.java
package com.example.a4;


import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;


import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;


public class Adaptor extends ListAdapter<DataClass,Adaptor.MyViewHolder> {
    Context context;

    Adaptor( Context context){
        super(DIFF_CALLBACK);
        this.context = context;
    }

    private static final DiffUtil.ItemCallback<DataClass> DIFF_CALLBACK = new DiffUtil.ItemCallback<DataClass>() {
        @Override
        public boolean areItemsTheSame(@NonNull DataClass oldItem, @NonNull DataClass newItem) {
            return oldItem.getId()==newItem.getId();
        }

        @Override
        public boolean areContentsTheSame(@NonNull DataClass oldItem, @NonNull DataClass newItem) {
            return oldItem.getNoteName().equals(newItem.getNoteName())&&
                    oldItem.getNote().equals(newItem.getNote());
        }
    };

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
      View itemView = LayoutInflater.from(context)
                .inflate(R.layout.cardview_adaptor,parent,false);
        return new MyViewHolder(itemView, context);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {
        holder.bind(getItem(position));
    }


     class MyViewHolder extends RecyclerView.ViewHolder  {
        TextView noteName,note,date_tv;
        Context context ;

         MyViewHolder(@NonNull View itemView, Context context) {
            super(itemView);
            this.context = context;
            noteName = itemView.findViewById(R.id.noteName_tv);
            note = itemView.findViewById(R.id.note_tv);
            date_tv = itemView.findViewById(R.id.date_tv);
        }
        void bind(final DataClass dataClass){
            noteName.setText(dataClass.getNoteName());
            note.setText(dataClass.getNote());
            date_tv.setText(dataClass.getDate());
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent intent = new Intent(context,NoteActivity.class);
                    intent.putExtra("id",dataClass.getId());
                    context.startActivity(intent);
                }
            });
            Log.d("tag4",dataClass.getNoteName());
        }

     }


}

File: app/src/main/java/com/example/a4/Adaptor.kt
package com.example.a4

import android.content.ContentResolver
import android.content.Context
import android.content.Intent
import android.database.Cursor
import android.net.Uri
import android.provider.MediaStore
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.a4.Adaptor.MyViewHolder
import kotlinx.android.synthetic.main.cardview_adaptor.view.*
import java.io.File

class Adaptor internal constructor(var context: Context) : ListAdapter<Notes, MyViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        val itemView = LayoutInflater.from(context)
                .inflate(R.layout.cardview_adaptor, parent, false)
        return MyViewHolder(itemView, context)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class MyViewHolder(itemView: View, var context: Context) : RecyclerView.ViewHolder(itemView) {
        fun bind(notesDataClass: Notes) {
            itemView.noteName_tv.text = notesDataClass.noteName
            itemView.note_tv.text = notesDataClass.note
            itemView.date_tv.text = notesDataClass.date
            if (notesDataClass.imageUri == null) itemView.noteImageView.visibility = View.GONE
            else {
                val cr: ContentResolver = context.contentResolver
                val projection = arrayOf(MediaStore.MediaColumns.DATA)
                var cur: Cursor? = cr.query(Uri.parse(notesDataClass.imageUri), projection, null, null, null)
                if (cur != null) {
                    if (cur.moveToFirst()) {
                        var filePath = cur.getString(0)

                        if (File(filePath).exists()) {
                            // do something if it exists
                            Glide.with(context).load(Uri.parse(notesDataClass.imageUri)).into(itemView.noteImageView)
                            itemView.noteImageView.visibility = View.VISIBLE
                        } else {
                            // File was not found
                            itemView.noteImageView.visibility = View.GONE
                        }
                    } else {
                        // Uri was ok but no entry found.
                        itemView.noteImageView.visibility = View.GONE
                    }
                    cur.close()
                } else {
                    // content Uri was invalid or some other error occurred
                    itemView.noteImageView.visibility = View.GONE
                }
            }


            itemView.setOnClickListener {
                val intent = Intent(context, NoteActivity::class.java)
                intent.putExtra("id", notesDataClass.id)
                context.startActivity(intent)
            }
        }

    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<Notes> = object : DiffUtil.ItemCallback<Notes>() {
            override fun areItemsTheSame(oldItem: Notes, newItem: Notes): Boolean {
                return oldItem.id === newItem.id
            }

            override fun areContentsTheSame(oldItem: Notes, newItem: Notes): Boolean {
                return oldItem.noteName == newItem.noteName && oldItem.note == newItem.note
            }
        }
    }

}